pkg/vpa/vpa.go

```diff
@@ -162,69 +155,126 @@ func (r Reconciler) namespaceIsManaged(namespace *corev1.Namespace) bool {
+func (r Reconciler) reconcileDeploymentsAndVPAs(nsName string, vpas []v1beta2.VerticalPodAutoscaler, deployments []appsv1.Deployment, vpaUpdateMode v1beta2.UpdateMode) error {
+	// these keys will eventually contain the leftover vpas that do not have a matching deployment associated
+	vpaHasAssociatedDeployment := map[string]bool{}
+	for _, deployment := range deployments {
+		var dvpa *v1beta2.VerticalPodAutoscaler
+		// search for the matching vpa (will have the same name)
+		for _, vpa := range vpas {
+			if deployment.Name == vpa.Name {
+				// found the vpa associated with this deployment
+				dvpa = &vpa
+				vpaHasAssociatedDeployment[dvpa.Name] = true
+				break
+			}
+		}
 
-	if len(vpaNeeded) == 0 {
-		klog.Info("All VPAs are in sync.")
-	} else if len(vpaNeeded) > 0 {
-		for _, vpaName := range vpaNeeded {
-			err := r.createVPA(nsName, vpaName, vpaUpdateMode)
+		// for logging
+		vpaName := "none"
+		if dvpa != nil {
+			vpaName = dvpa.Name
+		}
+		klog.V(2).Infof("Reconciling Namespace/%s for Deployment/%s with VPA/%s in vpa-update-mode=%s", nsName, deployment.Name, vpaName, vpaUpdateMode)
+		err := r.reconcileDeploymentAndVPA(nsName, deployment, dvpa, vpaUpdateMode)
+		if err != nil {
+			return err
+		}
+	}
+
+	for _, vpa := range vpas {
+		if _, ok := vpaHasAssociatedDeployment[vpa.Name]; !ok {
+			// these vpas do not have a matching deployment, delete them
+			klog.V(2).Infof("Deleting dangling VPA/%s in Namespace/%s", vpa.Name, nsName)
+			err := r.deleteVPA(nsName, vpa.Name)
 			if err != nil {
 				return err
 			}
 		}
 	}
 
-	// Now that this is one, we can delete any VPAs that don't have matching deployments.
-	vpaDelete := utils.Difference(vpaNames, deploymentNames)
-	klog.V(5).Infof("Diff vpas, deployments: %v", vpaDelete)
+	return nil
+}
+
+func (r Reconciler) reconcileDeploymentAndVPA(nsName string, deployment appsv1.Deployment, vpa *v1beta2.VerticalPodAutoscaler, vpaUpdateMode v1beta2.UpdateMode) error {
+	// check if the Deployment has its own vpa-update-mode set
+	if _, ok := deployment.GetAnnotations()[vpaUpdateModeKey]; ok {
+		vpaUpdateMode = vpaUpdateModeForResource(&deployment)
+		klog.V(5).Infof("Deployment/%s has custom vpa-update-mode=%s", deployment.Name, vpaUpdateMode)
+	}
 
-	if len(vpaDelete) == 0 {
-		klog.Info("No VPAs to delete.")
-	} else if len(vpaDelete) > 0 {
-		for _, vpaName := range vpaDelete {
-			err := r.deleteVPA(nsName, vpaName)
+	if vpa == nil {
+		klog.V(5).Infof("Deployment/%s does not have a VPA currently, creating VPA/%s", deployment.Name, deployment.Name)
+		// no vpa exists, create one (use the same name as the deployment)
+		err := r.createVPA(nsName, deployment.Name, vpaUpdateMode)
+		if err != nil {
+			return err
+		}
+	} else {
+		// vpa exists
+		if *vpa.Spec.UpdatePolicy.UpdateMode == vpaUpdateMode {
+			klog.V(5).Infof("Deployment/%s does have a VPA/%s and matches vpa-update-mode=%s", deployment.Name, vpa.Name, vpaUpdateMode)
```

Comment:
Maybe add something like `, leaving as-is` to the log, so it's obvious nothing is updating

Reply:
Same thing here, and I'll add a note to remember to update this wording in that PR.